pkg/webhook/preflight/nutanix/failuredomain.go

```diff
@@ -0,0 +1,211 @@
+				fdc.failureDomainName,
+				err,
+			),
+			Field: fdc.field,
+		})
+		return result
+	}
+
+	// Validate the failure domain configuration
+	// Validate spec.prismElementCluster configuration
+	peIdentifier := fdObj.Spec.PrismElementCluster
+	peClusters, err := getClusters(fdc.nclient, &peIdentifier)
+	if err != nil {
+		result.Allowed = false
+		result.InternalError = true
+		result.Causes = append(result.Causes, preflight.Cause{
+			Message: fmt.Sprintf(
+				"Failed to check if the Prism Element cluster %q, referenced by Failure Domain %q, exists: %v This is usually a temporary error. Please retry.", //nolint:lll // Message is long.
+				peIdentifier,
+				fdc.failureDomainName,
+				err,
+			),
+			Field: fdc.field,
+		})
+		return result
+	}
+	if len(peClusters) != 1 {
+		result.Allowed = false
+		result.Causes = append(result.Causes, preflight.Cause{
+			Message: fmt.Sprintf(
+				"Found %d Prism Element cluster(s) that match identifier %q. There must be exactly 1 that matches this identifier.", //nolint:lll // Message is long.
+				len(peClusters),
+				peIdentifier,
+			),
+			Field: fdc.field,
+		})
+		return result
+	}
+	peUUID := *peClusters[0].ExtId
+
+	// Validate spec.subnets configuration
+	for _, id := range fdObj.Spec.Subnets {
+		subnets, err := getSubnets(fdc.nclient, &id, peUUID)
+		if err != nil {
+			result.Allowed = false
+			result.InternalError = true
+			result.Causes = append(result.Causes, preflight.Cause{
+				Message: fmt.Sprintf(
+					"Failed to get subnet %q referenced by the Failure Domain %q: %v This is usually a temporary error. Please retry.", //nolint:lll // Message is long.
+					id,
+					fdc.failureDomainName,
+					err,
+				),
+				Field: fdc.field,
+			})
+			continue
+		}
+
+		if len(subnets) != 1 {
+			result.Allowed = false
+			result.Causes = append(result.Causes, preflight.Cause{
+				Message: fmt.Sprintf(
+					"Found %d Subnets that match identifier %q. There must be exactly 1 subnet that matches this identifier.", //nolint:lll // Message is long.,
+					len(subnets),
+					id,
+				),
+				Field: fdc.field,
+			})
+			continue
+		}
+	}
+
+	return result
+}
+
+// getSubnets returns the subnets found in PC with the input identifier and PE uuid.
+func getSubnets(client client, subnetId *capxv1.NutanixResourceIdentifier, peUUID string) ([]netv4.Subnet, error) {
+	switch {
+	case subnetId.IsUUID():
+		resp, err := client.GetSubnetById(subnetId.UUID)
+		if err != nil {
+			return nil, err
+		}
+		if resp == nil {
+			// No subnet returned.
+			return []netv4.Subnet{}, nil
+		}
+		subnet, ok := resp.GetData().(netv4.Subnet)
+		if !ok {
+			return nil, fmt.Errorf("failed to get data returned by GetSubnetById")
+		}
+		return []netv4.Subnet{subnet}, nil
+	case subnetId.IsName():
+		filter_ := fmt.Sprintf("name eq '%s' and clusterReference eq '%s'", *subnetId.Name, peUUID)
```

Comment:
what happens if the subnet is an overlay subnet?

Reply:
Good catch. Will fix it.